source/extensions/transport_sockets/tls/ssl_handshaker.cc

```diff
@@ -287,6 +287,16 @@ const std::string& SslHandshakerImpl::subjectPeerCertificate() const {
   return cached_subject_peer_certificate_;
 }
 
+const std::string& SslHandshakerImpl::requestedServerName() const {
```

Comment:
Should we populate the SNI into ConnectionInfoProvider? 
https://github.com/envoyproxy/envoy/blob/e7ebb0290d5527a002761d9d0aff51cb5f7c98cc/envoy/network/socket.h#L132-L135

Since tls inspect filter is doing the same thing. Otherwise, we will have multiple places to access the SNI.

Reply:
Hmm, I would actually prefer not use the plumbing from tls inspector to filer chain matcher, since I think it is rather convoluted and broken. The recent plumbing for ALPN in PR #18884 also did not use it.

I would prefer to use an abstraction that is clearly designed for the my purpose. Here are some issues I see with the connection info used by tls-inspector.
1. Some methods on core component interface are only returning values in the presence of an tls inspector extension. This is odd at the very least.
2. A few TLS related methods are glued to network connection interface which clearly abstracts L4 concerns.
3. The ownership of TLS properties by the Socket object further points to broken abstraction.
4. Access to TLS properties is weirdly split and some are duplicated between two interfaces Network::Connection and Ssl::Connection.

My take is that the tls-inspector to filter chain matcher plumbing should be re-implemented using metadata matcher and the TLS methods should be removed from L4 connection info interface. I can file an Issue for this refactor. It would pain me to use it in this way, as it has so many smells.

Reply:
You raise some valid points.

> Some methods on core component interface are only returning values in the presence of an tls inspector extension. This is odd at the very least.

This is because it doesn't have to be provided by TLS; a listener filter can be written for any other protocol that exposes a similar notion. Also, this data is available even if Envoy isn't terminating TLS, which is useful in some configurations.

> Access to TLS properties is weirdly split and some are duplicated between two interfaces Network::Connection and Ssl::Connection.

I agree; I would be in favor of cleaning this up. But the changes in this PR would further duplicate both API accessors and functionality for how the data is populated.

Also, there are some functional reasons for some of the split (but probably not all of it): serverName is available without TLS termination, but I think that Ssl::Connection is only available when we are doing TLS termination.

Logically, I agree that it would make sense to put serverName on Ssl::Connection. My only argument for not doing that is to avoid duplication, because it already exists somewhere else.

> My take is that the tls-inspector to filter chain matcher plumbing should be re-implemented using metadata matcher

That would be a nice cleanup, and would make it more general such that filter chain matching could match on anything that a listener filter wanted to produce. IIRC this code originated before we had metadata in Envoy; it's very old.